momepy/preprocessing.py

```diff
@@ -740,3 +745,368 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+            if edge[-1]["length"] > tolerance
+        ]
+    )
+    isolated_nodes_list = list(nx.isolates(components_graph))
+    components_graph.remove_nodes_from(isolated_nodes_list)
+
+    # The connected components of this graph are node clusters we must individually simplify. We collect them in a dataframe and
+    #  retrieve node properties (x, y coords mainly) from the original graph.
+    components_dict = dict(
+        enumerate(nx.connected_components(components_graph), start=max(nodes) + 1)
+    )
+    nodes_to_merge_dict = {
+        node: cpt for cpt, nodes in components_dict.items() for node in nodes
+    }
+    nodes_to_merge_df = pd.concat(
+        [
+            pd.DataFrame.from_dict(
+                nodes_to_merge_dict, orient="index", columns=["cluster"]
+            ),
+            nodes_df[[x_col, y_col]],
+        ],
+        axis=1,
+        join="inner",
+    )
+
+    # The two node attributes we need for the clusters are the position of the cluster centroids. Those are obtained by averaging
+    #  the x and y columns. We also add attribtues referring to the original osmids in every cluster:
+    cluster_centroids_df = nodes_to_merge_df.groupby("cluster").mean()
+    cluster_centroids_df["simplified"] = True
+    cluster_centroids_df["original_osmids"] = cluster_centroids_df.index.map(
+        components_dict
+    )
+    cluster_nodes_list = list(cluster_centroids_df.to_dict("index").items())
+
+    # Rebuild edges if necessary:
+    if rebuild_graph:
+        graph.graph["approach"] = "primal"
+        edges_gdf = mm.nx_to_gdf(graph, points=False, lines=True)
+        simplified_edges = _get_rebuilt_edges(
+            edges_gdf,
+            nodes_to_merge_dict,
+            cluster_centroids_df,
+            method=rebuild_edges_method,
+            tolerance=tolerance,
+            edge_from_col=edge_from_col,
+            edge_to_col=edge_to_col,
+        )
+
+    # Replacing the collapsed nodes with centroids and adding edges:
+    simplified_graph = graph.copy()
+    simplified_graph.remove_nodes_from(
+        nodes_to_merge_df.index
+    )  # removing nodes also removes edges!
+    simplified_graph.add_nodes_from(cluster_nodes_list)
+
+    if rebuild_graph:
+        simplified_graph.add_edges_from(simplified_edges)
+
+    return simplified_graph
+
+
+def _get_rebuilt_edges(
+    edges_gdf,
+    nodes_dict,
+    cluster_df,
+    method="linear",
+    tolerance=30,
+    edge_from_col="from",
+    edge_to_col="to",
+):
+    """
+    Rebuilds edge given (potential) new origin and destinations through a linear method: current line is linearly extended
+     to and from the new endpoints.
+
+    :param geometry: shapely.LineString, current edge
+    :param new_origin: shapely.Point or None, new origin point or None if origin remains the same
+    :param new_destination: shapely.Point or None, new destination point or None if destination remains the same
```

Comment:
Since you document this, can you follow the style above? For the sake of consistency.